lib/hanami/action/mime.rb

```diff
@@ -207,7 +211,14 @@ def format
       #     end
       #   end
       def content_type
```

Comment:
Thinking about this purely conceptually the `Content-Type` and response body format should be set based on a couple things according to the `Accepts` header section in the HTTP RFC.

## Scenarios

1. If the service supports one of the mime types specified in `Accepts` header it should respond with the most preferred based on the q scoring that also happens to be supported. See [RFC 2616](http://www.rfc-base.org/txt/rfc-2616.txt) Section 14.1 for details.
2. If the client specifies one or more mime types and the service does not support any of them then the service should respond with a 406 status code.
3. If the client does not specify any accept mime types then the service should default to its ideal mime type.

If I remember correctly the 406 status code is already handled for scenario 2 above, or at least partially handled. Maybe we should revisit that to make sure it is completely handled.

I am a little confused by the priority order of `get_common_mime_type_from_accept_header` and `get_custom_mime_type_from_accept_header`.

It seems in my mind based on the RFC if an accept header is provided with one or more acceptable mime types you have two possible scenarios.

1. the service doesn't support any of them and a 406 status code is responded. This already happens.
2. the service has been identified as supporting one or more of the mime types specified in the `Accepts` header and therefore needs to identified the out of the supported ones which is the most preferred based on the q score stuff and set the content type type to that.

I am not sure why we need `get_common_mime_type_from_accept_header` used at all when we know an accept header has been specified so we have to be in one of the above two states. Am I missing something?

It also seems instead of the `get_custom_mime_type_from_accept_header`, or maybe not instead, but maybe the inside of that method needs to be different. It seems it would need to be something like `best_q_match(accept, get_acceptable_mime_types(accept))`. That way you would be first getting the matching mime types specified by the client in the Accept header and then taking that resulting list and picking the most preferred one using the q score matching.